Reduce the number of allocation/reuse allocation of ndarray #1943

royxue opened the pull request:
Note about this PR before merging:
- This is the first step of gh-1943. If the Loop or LoopGC VM get used and allow_gc is False and no lazy operation, we will reuse some intermediate memory storage. We do this by making the storage_map of some variable point to the same list. We carefully check which variable can be shared. This make that some thunk will have there output available when they run.
- Currently this is done only for variable where ndim == 0. A follow up ticket will generalize this.

nouiz commented on theano/compile/tests/test_profiling.py:
```diff
@@ -11,94 +13,96 @@
-def test_profiling():
+class Test_profiling(unittest.TestCase):
+
+    def test_profiling(self):
```
Now there is a few test for profiling. Can you add a docstring with the goal of the tests. I think it would be useful.

nouiz commented on theano/gof/vm.py:
```diff
@@ -894,6 +894,68 @@ def make_all(self, profiler=None, input_storage=None,
+                        idx_v = dmap[idx_o]
+                        assert len(idx_v) == 1, "Here we only support the possibility to destroy one input"
+                        ins = node.inputs[idx_v[0]]
+                    if vmap and idx_o in vmap:
```
use "elif", small speed up.

nouiz replied:
forget that. This add an extra check.

nouiz commented on theano/gof/vm.py:
```diff
@@ -894,6 +894,68 @@ def make_all(self, profiler=None, input_storage=None,
+                        origin = view_of.get(ins, ins)
+                        view_of[out] = origin
+                        viewed_by[origin].append(out)
+                idx_o += 1                        
```
bad indentation.

nouiz commented on theano/gof/vm.py:
```diff
@@ -894,6 +894,68 @@ def make_all(self, profiler=None, input_storage=None,
+
+                for ins in node.inputs:
+                    assert not (ins in view_of and viewed_by[ins])
+                    if (ins.ndim == 0 and storage_map[ins][0] and ins not in fgraph.outputs and ins.owner):
```
why: storage_map[ins][0]?

nouiz replied:
Here you also need to check that ins won't be used again later. The dependencies dict in the profiling.py file line 807

RoyXue replied:
@nouiz  here, there is a strange error, I remembered it cannot find dependencies in the ProfileStats, so here i use a reallocated set to track the ins wont be used again

nouiz replied:
I suppose the strange error are fixed by storage_map[ins][0]? At this step, it always contant None. So this condition will always be False. You do not want that.

RoyXue replied:
Fred, here if the input variable is a constant, then it's storage_map[ins][0] has a value, I think this is what we can reallocate to outputs.

nouiz replied:
I forgot about constant.

But it is the reverse. We can't reuse the memory of Constant. Constant should not change. So you should do

storage_map[ins][0] is None.

Add a comment to explain this.

nouiz commented on theano/gof/vm.py:
```diff
@@ -894,6 +894,68 @@ def make_all(self, profiler=None, input_storage=None,
+                            # where gc
+                            for i in range(idx + 1, len(order)):
+                                for outs in order[i].outputs:
+                                    if outs.ndim == 0:
```
Also check that out isn't the view or destroyed version of one inputs.

nouiz commented on theano/gof/vm.py:
```diff
@@ -894,6 +894,68 @@ def make_all(self, profiler=None, input_storage=None,
+                                #where gc
+                                for i in range(idx + 1, len(order)):
+                                    for outs in order[i].outputs:
+                                        if outs.ndim == 0:
```
idem

nouiz commented on theano/gof/vm.py:
```diff
@@ -894,6 +894,68 @@ def make_all(self, profiler=None, input_storage=None,
+
+                        if reuse_outs:
+                            # if reusable output variable exists
+                            reallocated_ins.add(ins)
```
To help later code, I would do:

append((ins, reuse_outs)

that way you keep the possible case.

RoyXue replied:
@nouiz 
here, my origin idea is use this reallocated_ins as a if condition.
for a input variable if it's in reallocated_ins it means it had been reallocated, then we can skip this variable.

I didnt make sense of why we should also put reuse_outs info here.

nouiz replied:
I misunderstood reuse_outs.

nouiz commented on theano/gof/vm.py:
```diff
@@ -894,6 +894,68 @@ def make_all(self, profiler=None, input_storage=None,
+                            reallocated_ins.add(ins)
+
+
+        import pdb; pdb.set_trace()
```
Remove this line to know if travis tests do not crash with the current code.

RoyXue replied:
oh!!!  I forget to remove it before I push

nouiz commented on theano/gof/vm.py:
```diff
@@ -894,6 +894,68 @@ def make_all(self, profiler=None, input_storage=None,
+                                    for outs in order[i].outputs:
+                                        if outs.ndim == 0:
+                                            reuse_outs.append(outs)
+
```
To make this faster, you could break out of the 2 loops when you find 1 elements for reuse_outs. I do not have a quick way to do this. You could add after the inner loop this:

if reuse_outs:
   break

nouiz commented on theano/gof/vm.py:
```diff
@@ -152,7 +152,11 @@ def update_profile(self, profile):
+                profile.dependencies = self.dependencies.copy()
+            else:
+                profile.dependencies = {}
+
```
Why this case can happen? I think the old code was good. I think self.dependencies is always a Dict. When this isn't exact?

royxue commented:
@nouiz 
Hi, Fred, 
I this commit, first, I removed the if "self.allow_gc" condition, make the reallocated information is able to use when allow_gc is False.
Then I changed reallocated_info format. For now it is a dict, the key is input variable, the value is [storage_map[ins], [storage_map[outs]](this is a list including all available outs)]
The next part is in Loop/LoopGC, I added the reallocation part.

nouiz commented:
many tests on travis are failing

nouiz commented on theano/gof/vm.py:
```diff
@@ -167,6 +170,10 @@ class Loop(VM):
     """
 
+    def __init__(self, nodes, thunks, pre_call_clear, reallocated_info):
+        super(LoopGC, self).__init__(nodes, thunks, pre_call_clear)
```
LoopGC -> Loop

nouiz commented on theano/gof/vm.py:
```diff
@@ -179,6 +186,13 @@ def __call__(self):
+                        info = self.reallocated_info.get(ins, None)
+                        if info:
+                            for outs in info[1]:
+                                if not outs[0]:
```
Why this condition? If it is already allocated, why not still override it? I think we should overried it even if some space is already there. I'm not 100% sure, so what do you think? If we can do that, we won't need the loop, as we will always use the first element of info[1]

royxue replied:
@nouiz 
I think if this outs is allocated, then we skip it, until find an outs which havent been allocated. Is override available here?

nouiz commented:
There 2 error left in travis. I'll try to review more the code now.

nouiz commented on theano/gof/vm.py:
```diff
@@ -894,6 +935,77 @@ def make_all(self, profiler=None, input_storage=None,
+                assert not (ins in view_of and viewed_by[ins])
+                if (getattr(ins, 'ndim', None) == 0 and not storage_map[ins][0] 
+                    and ins not in fgraph.outputs and ins.owner 
+                    and dependencies.get(ins, None)):
```
Here you are missing a condition. Maybe there is still some computation that isn't done will use ins. You  need to add back a condition that was in the file profiling.py like this:

all([compute_map[v][0] for v in dependencies[ins]])):

Here you need to create a copy of compute_map for this and you need to update it after a node was executed.

nouiz commented on theano/gof/vm.py:
```diff
@@ -894,6 +935,77 @@ def make_all(self, profiler=None, input_storage=None,
+                            if reuse_outs:
+                                break
+                            for outs in order[i].outputs:
+                                if outs.ndim == 0 and outs not in viewed_by.get(ins, []):
```
"outs not in viewed_by.get(ins, [])" is always True due to the condition on line 982.
What you want to check is if this variable outs (rename it to out by the way), is a view of one inputs or a destroied version of one inputs. In that case, the preallocated output won't be used. So it is somithing like:
if outs.ndim == 0 and not order[i].op.view_map.get(out_index, []) and not order[i].op.destroy_map.get(out_index, []):

nouiz commented on theano/gof/vm.py:
```diff
@@ -894,6 +935,77 @@ def make_all(self, profiler=None, input_storage=None,
+
+                    if reuse_outs:
+                        # if reusable output variable exists
+                        reallocated_ins.add(ins)
```
reallocated_ins isn't used. I would get rid of it.

The current code could put in the same storage_map multiple input. You should create a pre_allocated set, where you put all the reuse_outs you got. Also, we only need 1 reuse_out per ins. So keep only one.

nouiz commented:
I think it advanced well. I have a few comments to finish before this part can be used.

royxue commented:
@nouiz 
The storage_map values are [None], how shall we check the value?

nouiz commented:
Use this mode:

```
            m = theano.compile.get_mode(theano.Mode(linker='vm_nogc'))
            m = m.excluding('fusion', 'inplace')

```
You can clean up the change to config.allow_gc, it do not work.

nouiz commented:
The fact that it need a rebase make travis not run anymore on this PR. So if you can make a rebase now, it would help development.

nouiz commented on theano/gof/vm.py:
```diff
@@ -889,6 +896,87 @@ def make_all(self, profiler=None, input_storage=None,
+                            if reuse_out:
+                                break
+                            for out in order[i].outputs:
+                                if getattr(out, 'ndim', None) == 0 and out not in pre_allocated:
```
You need to add an extra condition. I think this is what make tests fail.

add out.type == ins.type

nouiz commented:
There is test failing. I do not have the time to check. Can you do it?

royxue commented:
@nouiz 
Is this can be merged now?

nouiz commented:
theano.sandbox.cuda.tests.test_basic_ops.T_Join_and_Split
theano/gof/tests/test_vm.py

If I run Theano tests like this:

THEANO_FLAGS=linker=vm theano-nose

I get some errors (It didn't finish running). Here is the 3 errors I have up to now. You can reproduce them with this:

THEANO_FLAGS=linker=vm theano-nose theano/gof/tests/test_vm.py:test_allow_gc_cvm theano/sandbox/cuda/tests/test_basic_ops.py:T_Join_and_Split.test_stack_hessian2 theano/sandbox/cuda/tests/test_basic_ops.py:T_subtensor.test_shape_i_const

Can you look into them and fix them? We need all tests to pass with the first command before we can merge. This trigger extra check, as by default it do not trigger the usage of your new functionality.

royxue commented:
Sure, I will fix this.
On Sat, Feb 28, 2015 at 12:33 AM Frédéric Bastien <notifications@github.com>
wrote:

> theano.sandbox.cuda.tests.test_basic_ops.T_Join_and_Split
> theano/gof/tests/test_vm.py
>
> If I run Theano tests like this:
>
> THEANO_FLAGS=linker=vm theano-nose
>
> I get some errors (It didn't finish running). Here is the 3 errors I have
> up to now. You can reproduce them with this:
>
> THEANO_FLAGS=linker=vm theano-nose
> theano/gof/tests/test_vm.py:test_allow_gc_cvm
> theano/sandbox/cuda/tests/test_basic_ops.py:T_Join_and_Split.test_stack_hessian2
> theano/sandbox/cuda/tests/test_basic_ops.py:T_subtensor.test_shape_i_const
>
> Can you look into them and fix them? We need all tests to pass with the
> first command before we can merge. This trigger extra check, as by default
> it do not trigger the usage of your new functionality.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/2287#issuecomment-76424149>.
>

nouiz commented:
Do you have some news? I'll be away next week for a conference. I probably can answer question during that time, but not sure how frequently.

royxue commented:
@nouiz 
No, kinda busy last week, but I will do something this week, fix the bugs errors first.

nouiz commented:
ok. thanks

On Tue, Mar 10, 2015 at 9:05 PM, Roy Xue <notifications@github.com> wrote:

> @nouiz <https://github.com/nouiz>
> No, kinda busy last week, but I will do something this week, fix the bugs
> errors first.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/2287#issuecomment-78181771>.
>

royxue commented:
@nouiz 
for this test this line
theano/gof/tests/test_vm.py:test_allow_gc_cvm 
assert f.fn.storage_map[n][0] is None
i think it is the old test file, now, what we do is to change the storage_map, so for this, I think maybe we should delete this line

nouiz commented:
I do not think so. The function tested have only 1 output and only 1 node.

This test is currently failing with linker=vm and Theano master. So it isn't a problem your PR introduced. If you have the time, can you look into it? Otherwise, make a ticket so we do not forget this case.

nouiz commented:
We want to make a new Theano release Thursday. I'm not sure I want to merge now this PR as this is stuff that is hard to review without forgetting corner case. So I'll wait after Theano 0.7 to merge this.

But you can start a new PR already that depend on this PR. I also want to make a new review pass before merging this. Just in case.
